UpdateSettingsIT#testEngineGCDeletesSetting failure

test failing:

```
org.elasticsearch.indices.settings.UpdateSettingsIT testEngineGCDeletesSetting
```

#38254 introduced a new `VersionConflictEngineException` assertion that seems 
to be dependent on cache invalidation timed by `Thread.sleep`.

exception:

```
java.lang.AssertionError: expected a class org.elasticsearch.index.engine.VersionConflictEngineException exception to be thrown

	at __randomizedtesting.SeedInfo.seed([103812063D7BAB96:96D53E0C3D9BE721]:0)
	at org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertThrows(ElasticsearchAssertions.java:596)
	at org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertThrows(ElasticsearchAssertions.java:543)
	at org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertThrows(ElasticsearchAssertions.java:520)
	at org.elasticsearch.indices.settings.UpdateSettingsIT.testEngineGCDeletesSetting(UpdateSettingsIT.java:452)
```

was tough to get a stacktrace, but looks to fail in `InternalEngine#planIndexingAsPrimary`

with the exception:

```[type][1]: version conflict, required seqNo [3], primary term [1]. but no document was found```

ci link: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+7.0+internalClusterTest/472/console


reproduce:

```
./gradlew :server:integTest \
  -Dtests.seed=EAF25E026AF61CCF \
  -Dtests.class=org.elasticsearch.indices.settings.UpdateSettingsIT \
  -Dtests.method="testEngineGCDeletesSetting" \
  -Dtests.security.manager=true \
  -Dtests.locale=is-IS \
  -Dtests.timezone=Australia/Canberra \
  -Dcompiler.java=11 \
  -Druntime.java=8
```

This is easy to reproduce if one removes the `Thread.sleep(300)`. Which means that
cache hadn't changed in the CI run in time for the assertion.

--- a/server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java
+++ b/server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java
@@ -550,22 +550,39 @@ public class ThreadPool implements Scheduler, Closeable {
             this.absoluteMillis = System.currentTimeMillis();
             this.counter = new TimeCounter();
             setDaemon(true);
+            if (interval <= 0) {
+                this.running = false;
+            }
         }
 
         /**
          * Return the current time used for relative calculations. This is
          * {@link System#nanoTime()} truncated to milliseconds.
+         * <p>
+         * If {@link ThreadPool#ESTIMATED_TIME_INTERVAL_SETTING} is set to 0
+         * then the cache is disabled and the method calls {@link System#nanoTime()}
+         * whenever called. Typically used for testing.
          */
         long relativeTimeInMillis() {
-            return relativeMillis;
+            if (running) {
+                return relativeMillis;
+            }
+            return TimeValue.nsecToMSec(System.nanoTime());
         }
 
         /**
          * Return the current epoch time, used to find absolute time. This is
          * a cached version of {@link System#currentTimeMillis()}.
+         * <p>
+         * If {@link ThreadPool#ESTIMATED_TIME_INTERVAL_SETTING} is set to 0
+         * then the cache is disabled and the method calls {@link System#currentTimeMillis()}
+         * whenever called. Typically used for testing.
          */
         long absoluteTimeInMillis() {
-            return absoluteMillis;
+            if (running) {
+                return absoluteMillis;
+            }
+            return System.currentTimeMillis();
         }
 
         @Override

--- a/server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java
+++ b/server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java
@@ -550,9 +550,6 @@ public class ThreadPool implements Scheduler, Closeable {
             this.absoluteMillis = System.currentTimeMillis();
             this.counter = new TimeCounter();
             setDaemon(true);
-            if (interval <= 0) {
-                this.running = false;
-            }
         }
 
         /**
@@ -564,7 +561,7 @@ public class ThreadPool implements Scheduler, Closeable {
          * whenever called. Typically used for testing.
          */
         long relativeTimeInMillis() {
-            if (running) {
+            if (0 < interval) {
                 return relativeMillis;
             }
             return TimeValue.nsecToMSec(System.nanoTime());
@@ -579,7 +576,7 @@ public class ThreadPool implements Scheduler, Closeable {
          * whenever called. Typically used for testing.
          */
         long absoluteTimeInMillis() {
-            if (running) {
+            if (0 < interval) {
                 return absoluteMillis;
             }
             return System.currentTimeMillis();
@@ -587,7 +584,7 @@ public class ThreadPool implements Scheduler, Closeable {
 
         @Override
         public void run() {
-            while (running) {
+            while (running && 0 < interval) {
                 relativeMillis = TimeValue.nsecToMSec(System.nanoTime());
                 absoluteMillis = System.currentTimeMillis();
                 try {

--- a/server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java
+++ b/server/src/main/java/org/elasticsearch/threadpool/ThreadPool.java
@@ -162,7 +162,8 @@ public class ThreadPool implements Scheduler, Closeable {
     }
 
     public static Setting<TimeValue> ESTIMATED_TIME_INTERVAL_SETTING =
-        Setting.timeSetting("thread_pool.estimated_time_interval", TimeValue.timeValueMillis(200), Setting.Property.NodeScope);
+        Setting.timeSetting("thread_pool.estimated_time_interval",
+            TimeValue.timeValueMillis(200), TimeValue.ZERO, Setting.Property.NodeScope);
 
     public ThreadPool(final Settings settings, final ExecutorBuilder<?>... customBuilders) {
         assert Node.NODE_NAME_SETTING.exists(settings);

--- a/server/src/test/java/org/elasticsearch/indices/settings/UpdateSettingsIT.java
+++ b/server/src/test/java/org/elasticsearch/indices/settings/UpdateSettingsIT.java
@@ -26,7 +26,6 @@ import org.elasticsearch.cluster.metadata.IndexMetaData;
 import org.elasticsearch.common.Priority;
 import org.elasticsearch.common.settings.Setting;
 import org.elasticsearch.common.settings.Settings;
-import org.elasticsearch.common.unit.TimeValue;
 import org.elasticsearch.index.IndexModule;
 import org.elasticsearch.index.IndexService;
 import org.elasticsearch.index.engine.VersionConflictEngineException;

--- a/server/src/test/java/org/elasticsearch/indices/settings/UpdateSettingsIT.java
+++ b/server/src/test/java/org/elasticsearch/indices/settings/UpdateSettingsIT.java
@@ -26,12 +26,14 @@ import org.elasticsearch.cluster.metadata.IndexMetaData;
 import org.elasticsearch.common.Priority;
 import org.elasticsearch.common.settings.Setting;
 import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.unit.TimeValue;
 import org.elasticsearch.index.IndexModule;
 import org.elasticsearch.index.IndexService;
 import org.elasticsearch.index.engine.VersionConflictEngineException;
 import org.elasticsearch.indices.IndicesService;
 import org.elasticsearch.plugins.Plugin;
 import org.elasticsearch.test.ESIntegTestCase;
+import org.elasticsearch.threadpool.ThreadPool;
 
 import java.util.Arrays;
 import java.util.Collection;
@@ -124,6 +126,16 @@ public class UpdateSettingsIT extends ESIntegTestCase {
         public List<Setting<?>> getSettings() {
             return Collections.singletonList(FINAL_SETTING);
         }
+    }
+
+    /**
+     * Needed by {@link UpdateSettingsIT#testEngineGCDeletesSetting()}
+     */
+    @Override
+    protected Settings nodeSettings(int nodeOrdinal) {
+        return Settings.builder().put(super.nodeSettings(nodeOrdinal))
+            .put("thread_pool.estimated_time_interval", TimeValue.timeValueMillis(1))
+            .build();
     }
 
     public void testUpdateDependentClusterSettings() {
@@ -435,7 +447,7 @@ public class UpdateSettingsIT extends ESIntegTestCase {
         assertThat(getSettingsResponse.getSetting("test", "index.final"), nullValue());
     }
 
-    public void testEngineGCDeletesSetting() throws InterruptedException {
+    public void testEngineGCDeletesSetting() {
         createIndex("test");
         client().prepareIndex("test", "type", "1").setSource("f", 1).get();
         DeleteResponse response = client().prepareDelete("test", "type", "1").get();
@@ -443,15 +455,23 @@ public class UpdateSettingsIT extends ESIntegTestCase {
         long primaryTerm = response.getPrimaryTerm();
         // delete is still in cache this should work
         client().prepareIndex("test", "type", "1").setSource("f", 2).setIfSeqNo(seqNo).setIfPrimaryTerm(primaryTerm).get();
-        client().admin().indices().prepareUpdateSettings("test").setSettings(Settings.builder().put("index.gc_deletes", 0)).get();
+        assertAcked(client().admin().indices().prepareUpdateSettings("test").setSettings(Settings.builder().put("index.gc_deletes", 0)));
 
         response = client().prepareDelete("test", "type", "1").get();
         seqNo = response.getSeqNo();
-        Thread.sleep(300); // wait for cache time to change TODO: this needs to be solved better. To be discussed.
+
+        // Make sure the time has advanced for InternalEngine#resolveDocVersion()
+        for (ThreadPool tPool : internalCluster().getInstances(ThreadPool.class)) {
+            long time1 = tPool.relativeTimeInMillis();
+            long time2 = tPool.relativeTimeInMillis();
+            while (time1 == time2) {
+                time2 = tPool.relativeTimeInMillis();
+            }
+        }
+
         // delete is should not be in cache
         assertThrows(client().prepareIndex("test", "type", "1").setSource("f", 3).setIfSeqNo(seqNo).setIfPrimaryTerm(primaryTerm),
             VersionConflictEngineException.class);
-
     }
 
     public void testUpdateSettingsWithBlocks() {

--- a/server/src/test/java/org/elasticsearch/indices/settings/UpdateSettingsIT.java
+++ b/server/src/test/java/org/elasticsearch/indices/settings/UpdateSettingsIT.java
@@ -461,9 +461,9 @@ public class UpdateSettingsIT extends ESIntegTestCase {
         seqNo = response.getSeqNo();
 
         // Make sure the time has advanced for InternalEngine#resolveDocVersion()
-        for (ThreadPool tPool : internalCluster().getInstances(ThreadPool.class)) {
-            long time1 = tPool.relativeTimeInMillis();
-            assertBusy(() -> assertThat(tPool.relativeTimeInMillis(), greaterThan(time1)));
+        for (ThreadPool threadPool : internalCluster().getInstances(ThreadPool.class)) {
+            long startTime = threadPool.relativeTimeInMillis();
+            assertBusy(() -> assertThat(threadPool.relativeTimeInMillis(), greaterThan(startTime)));
         }
 
         // delete is should not be in cache

--- a/server/src/test/java/org/elasticsearch/indices/settings/UpdateSettingsIT.java
+++ b/server/src/test/java/org/elasticsearch/indices/settings/UpdateSettingsIT.java
@@ -48,6 +48,7 @@ import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertBloc
 import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertThrows;
 import static org.hamcrest.Matchers.containsString;
 import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.greaterThan;
 import static org.hamcrest.Matchers.nullValue;
 
 public class UpdateSettingsIT extends ESIntegTestCase {
@@ -446,7 +447,7 @@ public class UpdateSettingsIT extends ESIntegTestCase {
         assertThat(getSettingsResponse.getSetting("test", "index.final"), nullValue());
     }
 
-    public void testEngineGCDeletesSetting() {
+    public void testEngineGCDeletesSetting() throws Exception {
         createIndex("test");
         client().prepareIndex("test", "type", "1").setSource("f", 1).get();
         DeleteResponse response = client().prepareDelete("test", "type", "1").get();
@@ -462,10 +463,7 @@ public class UpdateSettingsIT extends ESIntegTestCase {
         // Make sure the time has advanced for InternalEngine#resolveDocVersion()
         for (ThreadPool tPool : internalCluster().getInstances(ThreadPool.class)) {
             long time1 = tPool.relativeTimeInMillis();
-            long time2 = tPool.relativeTimeInMillis();
-            while (time1 == time2) {
-                time2 = tPool.relativeTimeInMillis();
-            }
+            assertBusy(() -> assertThat(tPool.relativeTimeInMillis(), greaterThan(time1)));
         }
 
         // delete is should not be in cache

--- a/server/src/test/java/org/elasticsearch/indices/settings/UpdateSettingsIT.java
+++ b/server/src/test/java/org/elasticsearch/indices/settings/UpdateSettingsIT.java
@@ -134,7 +134,7 @@ public class UpdateSettingsIT extends ESIntegTestCase {
     @Override
     protected Settings nodeSettings(int nodeOrdinal) {
         return Settings.builder().put(super.nodeSettings(nodeOrdinal))
-            .put("thread_pool.estimated_time_interval", TimeValue.timeValueMillis(1))
+            .put("thread_pool.estimated_time_interval", 0)
             .build();
     }
 

--- a/server/src/test/java/org/elasticsearch/threadpool/ThreadPoolTests.java
+++ b/server/src/test/java/org/elasticsearch/threadpool/ThreadPoolTests.java
@@ -19,8 +19,10 @@
 
 package org.elasticsearch.threadpool;
 
+import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.test.ESTestCase;
 
+import static org.elasticsearch.threadpool.ThreadPool.ESTIMATED_TIME_INTERVAL_SETTING;
 import static org.hamcrest.CoreMatchers.equalTo;
 
 public class ThreadPoolTests extends ESTestCase {
@@ -58,5 +60,11 @@ public class ThreadPoolTests extends ESTestCase {
             threadPool.shutdown();
             threadPool.close();
         }
+    }
+
+    public void testEstimatedTimeIntervalSettingAcceptsOnlyZeroAndPositiveTime() {
+        Settings settings = Settings.builder().put("thread_pool.estimated_time_interval", -1).build();
+        Exception e = expectThrows(IllegalArgumentException.class, () -> ESTIMATED_TIME_INTERVAL_SETTING.get(settings));
+        assertEquals("failed to parse value [-1] for setting [thread_pool.estimated_time_interval], must be >= [0ms]", e.getMessage());
     }
 }

